### PR TITLE
Refactors get_snapshot_storages()

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

`AccountsDb::get_snapshot_storages()` is due for a refactor. We can speed it up for the common case, and also simplify.

Since #3737, we now call `get_snapshot_storages()` every time we `clean`. The observation here is we'll only need about 100 storages (on average), yet the current impl of `get_snapshot_storages()` Arc::clone's *all* the storages, and then filters out the unneeded ones. We can change this to only Arc::clone the useful ones instead.

Additionally, the filter step is done in parallel. When we only have 100 storages, the parallel execution does not help. In fact, with a chunk size of 5000, we end up getting zero benefit, but do have to pay the cost of running in the thread pool.

#### Summary of Changes

* When getting storages, only Arc::clone the ones we need
* When filtering, do not use a thread pool

#### Results

I ran this on against mnb and saw good results.

Since `get_snapshot_storages()` is called in two places, I wanted to look at the perf results in both.

1. `clean`

Here, we only need ~100 storages each time. So not Arc::cloning and not using the thread pool really helps. The PR ends up running consistently, and meaningfully, faster:

| branch | get storages | filter | total |
|--------|--------|--------|---------|
| master | 30-50 ms | 100-400 us | 30-50 ms |
| pr v1 | 11-13 ms | 10-30 us | 11-13 ms |
| pr v2 | n/a | n/a | 7-9 ms |
| pr v3 | n/a | n/a | 6-8 ms |

2. taking full snapshots

Full snapshots do need all the storages, so we will end up Arc::cloning almost all the storages. And it's possible the thread pool does help here. For the most part, runtimes are pretty similar. The PR does have a worse worst-case filter time.

| branch | get storages | filter | total |
|--------|--------|--------|--------|
| master | 30-50 ms | 30-40 ms  | 60-90 ms |
| pr v1 | 30-50 ms | 30-50 ms  | 60-100 ms |
| pr v2 | n/a | n/a | 40-60 ms |
| pr v3 | n/a | n/a | 40-60 ms |

Overall, I think the common case of `clean` makes this change clearly a win. There is maybe a slight slow down for full snapshots, but that code is both infrequent, and in the background, so I don't think it matters much. Additionally, by not using a thread pool, we may reduce resource usage for the system as a whole.

### CODE REVIEW

**Comment by jeffwashington** [comment]
this is fine. But, since it is now called very clean, another alternative is to get all roots less than the cutoff and then get storages based on roots. I imagine that's faster by a lot. Or, we just leave this like it is since by the time master is rolled out we are hopefully ready to skip rewrites finally and enable ancient packing. Not sure what to do.  If you're not backporting this, I'd just leave it alone I imagine.

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -8473,44 +8473,21 @@ impl AccountsDb {
...
-                        .filter_map(|(slot, store)| {
-                            let store = std::mem::take(store).unwrap();
-                            store.has_accounts().then_some((store, *slot))
-                        })
-                        .collect::<Vec<(Arc<AccountStorageEntry>, Slot)>>()
-                })
-                .flatten()
+                .into_vec()
+                .into_iter()
+                .filter(|(slot, _)| self.accounts_index.is_alive_root(*slot))
```
Comment (HaoranYi) [inline accounts-db/src/accounts_db.rs]: Is it possible to move this alive root check into get-if and unzip them in get-if without creating the temporary slot_storage vec?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs]: Done!

**Comment by brooksprumo** [comment]
> since it is now called very clean, another alternative is to get all roots less than the cutoff and then get storages based on roots. I imagine that's faster by a lot. 

Yes, this does make an improvement for both clean and snapshots. Thanks for the suggestion!

> Or, we just leave this like it is since by the time master is rolled out we are hopefully ready to skip rewrites finally and enable ancient packing. Not sure what to do. If you're not backporting this, I'd just leave it alone I imagine.

My initial idea was to backport to v2.1 but not v2.0. Since we're planning to activate skipping rewrites in v2.1, I want to reduce the negative impact of getting the snapshot storages in every iteration of `clean`. *Not* backporting to v2.1 will still help with snapshots, but not with clean.

File: accounts-db/src/accounts_db.rs (hunk 1 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: Thanks! lgtm.

**Comment by HaoranYi** [comment]
looks like ci is failing ...

Not sure if it is because of the pr or ci flakey 

I will approve when ci passed.

**Comment by brooksprumo** [comment]
> looks like ci is failing ...
> 
> Not sure if it is because of the pr or ci flakey
> 
> I will approve when ci passed.

Here's some discussion from Discord on the CI issue: https://discord.com/channels/428295358100013066/560503042458517505/1310571536221995068

We'll need PR #3774 to land, and then I can rebase this PR to get the CI fix.

**Comment by brooksprumo** [comment]
PR #3774 was merged. I've rebased and force-pushed to pull in this fix. No code was changed.

**Review by jeffwashington — APPROVED** [review]
lgtm

**Comment by brooksprumo** [comment]
Merging! @HaoranYi, I'm interpreting https://github.com/anza-xyz/agave/pull/3760#issuecomment-2496127118 as an approval. Lmk if that's wrong!

**Comment by brooksprumo** [comment]
@jeffwashington @HaoranYi I'd like to backport to v2.1, since 2.1 and 2.0 are the main ones that'll be impacted by cleaning old storages during `clean`. Any objections?

**Comment by HaoranYi** [comment]
> @jeffwashington @HaoranYi I'd like to backport to v2.1, since 2.1 and 2.0 are the main ones that'll be impacted by cleaning old storages during `clean`. Any objections?

Fine with me. Do you happen to have the updated benchmark time after the new changes?

**Comment by brooksprumo** [comment]
> Do you happen to have the updated benchmark time after the new changes?

The v3 numbers correspond to the code that was merged. Are those the numbers you're looking for?

**Comment by HaoranYi** [comment]
yeah. lgtm.

**Pull request merged**